Support Nuvoton target NUMAKER_M252KG

ccli8 opened the pull request:
### Description

This PR adds Nuvoton's target NUMAKER_M252KG. NUMAKER_M252KG is Cortex-M23 based **without TrustZone**. It supports the following toolchains:

- ARMC6
- GCC_ARM
- IAR

### Pull request type

    [ ] Fix
    [ ] Refactor
    [x] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

ccli8 committed:
Support Nuvoton's NUMAKER_M252KG target

ccli8 committed:
[M252KG] Enlarge required deep sleep latency

This configuration is to pass wake-up from deep-sleep test such as mbedmicro-rtos-mbed-systimer.

ccli8 committed:
[M252KG] Override mpu-rom-end to 0x1fffffff

Without this override, mpu hal will require 5 mpu regions which exceed 4 mpu
regions supported by M252 (see hal/mpu/mbed_mpu_v8m.c). In this scenario,
we will hit assert error but we actually meet stack overrun first due to just
0x400 bytes for emitting error message. The issue doesn’t occur on other
targets such as M487 because it has 8 mpu regions.

ccli8 committed:
[M252KG] Enlarge LPTICKER_DELAY_TICKS for safe

On Nuvoton targets, lp_ticker_set_interrupt(...) needs around 3 lp-ticker
ticks to take effect. It may miss when current tick and match tick are very
close (see hal/LowPowerTickerWrapper.cpp). Enlarge LPTICKER_DELAY_TICKS to
4 from 3 to address this boundary case.

ccli8 committed:
[M252KG] Force enum PinName to 32-bit

NU_PINNAME_BIND(...) requires enum PinName to be 32-bit to encode module
binding information in it.

ccli8 committed:
[M252KG] Exclude USB UART from testing

USB UART is dedicated to USB COM and so must exclude from FPGA CI testing.

ccli8 committed:
[M252KG] Fix IP initialization sequence

Better IP initialization sequence:
1. Configure IP pins
2. Select IP clock source and then enable it
3. Reset the IP (SYS_ResetModule)

NOTE1: IP reset takes effect regardless of IP clock. So it doesn't matter if
       IP clock enable is before IP reset.
NOTE2: Non-configured pins may disturb IP's state, so IP pinout first and then
       IP reset.
NOTE3: IP reset at the end of IP initialization sequence can cover unexpected
       situation.

ccli8 committed:
[M252KG] Fix I2C NACK error

Fix logic error on replying NACK at the end of transfer.

This is also to fix FPGA CI test mbed_hal_fpga_ci_test_shield-i2c/
i2c - test single byte read i2c API.

ccli8 committed:
[M252KG] Fix redundant SPI clock generation

Fix SPI clocks are generated redundantly at the end of transfer.

This is also to fix FPGA CI test mbed_hal_fpga_ci_test_shield-spi/
SPI - async mode.

ccli8 committed:
[M252KG] Fix redundant call to UART IRQ handler

Honor RxIrq/TxIrq to avoid redundant call to UART IRQ handler.

This is also to fix FPGA CI test mbed_hal_fpga_ci_test_shield-uart.

ccli8 committed:
[M252KG] Support GPIO input pull-high/pull-low

In Nuvoton, only new-design chips support GPIO input pull-high/pull-low modes.
Targets not supporting this feature are listed below:

- NUMAKER_PFM_NANO130
- NUMAKER_PFM_NUC472
- NUMAKER_PFM_M453

ccli8 committed:
[M252KG] Free up peripheral pins in peripheral free-up HAL API

Without free-up of peripheral pins, peripheral pins of the same peripheral may
share by multiple ports after port iteration, and this peripheral may fail with
pin interference.

VeijoPesonen commented on features/storage/TESTS/kvstore/static_tests/main.cpp:
```diff
@@ -31,13 +31,20 @@ static char         buffer[20] = {};
 static const size_t data_size = 5;
 static size_t       actual_size = 0;
 static const size_t buffer_size = 20;
+
```
Having target specific settings inside generic test cases isn't something we can allow. What kind of issue you're trying to solve here?

ccli8 replied:
NUMAKER_M252KG just has 32KiB SRAM and meet OOM error in the `kvstore-static_tests` test. So I decrease the number of forked threads from 3 to 2 to reduce memory footprint. I will send another PR to address this issue separately.

ccli8 committed:
[M252KG] Add 'sectors' target configuration parameter

0xc0170 commented on targets/TARGET_NUVOTON/TARGET_M251/device/StdDriver/inc/m251_scuart.h:
```diff
@@ -0,0 +1,266 @@
+ * @brief    M251 Smartcard UART mode (SCUART) driver header file
+ *
+ * @copyright (C) 2018 Nuvoton Technology Corp. All rights reserved.
+ *****************************************************************************/
```
Please add the license information (including SPDX identifier) to all StdDriver - already StdDriver for some targets contain the license

ccli8 replied:
@0xc0170 Fixed by adding **BSD-3-Clause** license claim to these BSP files

0xc0170 commented on targets/TARGET_NUVOTON/TARGET_M251/trng_api.c:
```diff
@@ -0,0 +1,90 @@
+
+#define PRNG_KEY_SIZE  (0x20UL)
+
+static void trng_get(unsigned char *pConversionData)
```
@ARMmbed/mbed-os-crypto Please review

ccli8 committed:
[M252KG] Add BSD-3-Clause license for BSP files

ccli8 committed:
[M252KG] Add missing 'sectors' into arm_pack_manager/index.json

ccli8 committed:
[M252KG] Remove TRNG support

Reasons to remove TRNG support:
1.  M252 just has 32KiB SRAM and cannot afford mbedtls application.
2.  Implementing TRNG HAL with PRNG H/W has security concern.

yanesca reviewed (approved):
LGTM.
